### PR TITLE
Update to Vert.x 4.4.1

### PR DESCRIPTION
**Pull request by cescoffier**
- Update Netty to version 4.1.90.Final
- Update Vert.x to version 4.4.1
- Update Mutiny bindings to version 3.3.0
- Update Quarkus HTTP to version 5.0.2.Final

### CODE REVIEW

**Comment by geoand** [comment]
CI seems to not like it:

```posh
2023-04-05T06:11:49.7367657Z [ERROR] Non-resolvable import POM: Could not find artifact io.smallrye.reactive:vertx-mutiny-clients-bom:pom:3.3.0-SNAPSHOT @ io.quarkus:quarkus-bom:999-SNAPSHOT, /home/runner/work/quarkus/quarkus/bom/application/pom.xml, line 414, column 25
2023-04-05T06:11:49.7368862Z [ERROR] Non-resolvable import POM: Could not find artifact io.smallrye.reactive:vertx-mutiny-clients-bom:pom:3.3.0-SNAPSHOT @ io.quarkus:quarkus-bom:999-SNAPSHOT, /home/runner/work/quarkus/quarkus/bom/application/pom.xml, line 414, column 25
2023-04-05T06:11:49.7369372Z 
2023-04-05T06:11:49.7369714Z     at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:397)
2023-04-05T06:11:49.7370407Z     at org.apache.maven.graph.DefaultGraphBuilder.collectProjects (DefaultGraphBuilder.java:414)
2023-04-05T06:11:49.7371198Z     at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor (DefaultGraphBuilder.java:405)
2023-04-05T06:11:49.7371918Z     at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:82)
2023-04-05T06:11:49.7380777Z     at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:535)
2023-04-05T06:11:49.7384110Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:220)
2023-04-05T06:11:49.7387066Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:193)
2023-04-05T06:11:49.7389947Z     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:106)
2023-04-05T06:11:49.7393861Z     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:963)
2023-04-05T06:11:49.7401227Z     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:296)
2023-04-05T06:11:49.7404343Z     at org.apache.maven.cli.MavenCli.main (MavenCli.java:199)
2023-04-05T06:11:49.7407459Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
2023-04-05T06:11:49.7410569Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
2023-04-05T06:11:49.7417637Z     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
2023-04-05T06:11:49.7423180Z     at java.lang.reflect.Method.invoke (Method.java:566)
2023-04-05T06:11:49.7426797Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
2023-04-05T06:11:49.7430851Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
2023-04-05T06:11:49.7434026Z     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
2023-04-05T06:11:49.7438679Z     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
2023-04-05T06:11:49.7441435Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
2023-04-05T06:11:49.7444946Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
2023-04-05T06:11:49.7458098Z     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
2023-04-05T06:11:49.7460502Z     at java.lang.reflect.Method.invoke (Method.java:566)
2023-04-05T06:11:49.7462712Z     at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:52)
2023-04-05T06:11:49.7464952Z     at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:161)
2023-04-05T06:11:49.7470901Z     at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:73)
2023-04-05T06:11:49.7473251Z [ERROR]   
2023-04-05T06:11:49.7475187Z [ERROR]   The project io.quarkus:quarkus-bom:999-SNAPSHOT (/home/runner/work/quarkus/quarkus/bom/application/pom.xml) has 1 error
2023-04-05T06:11:49.7484158Z [ERROR]     Non-resolvable import POM: Could not find artifact io.smallrye.reactive:vertx-mutiny-clients-bom:pom:3.3.0-SNAPSHOT @ line 414, column 25 -> [Help 2]
2023-04-05T06:11:49.7487147Z org.apache.maven.model.resolution.UnresolvableModelException: Could not find artifact io.smallrye.reactive:vertx-mutiny-clients-bom:pom:3.3.0-SNAPSHOT
```

File: bom/application/pom.xml [hunk 1]
```diff
@@ -68,7 +68,7 @@
         <smallrye-context-propagation.version>2.1.0</smallrye-context-propagation.version>
         <smallrye-reactive-streams-operators.version>1.0.13</smallrye-reactive-streams-operators.version>
         <smallrye-reactive-types-converter.version>3.0.0</smallrye-reactive-types-converter.version>
-        <smallrye-mutiny-vertx-binding.version>3.2.0</smallrye-mutiny-vertx-binding.version>
+        <smallrye-mutiny-vertx-binding.version>3.3.0-SNAPSHOT</smallrye-mutiny-vertx-binding.version>
```
Comment (geoand) [inline bom/application/pom.xml:71]: I assume this is an oversight :)

File: integration-tests/grpc-interceptors/pom.xml [hunk 2]
```diff
@@ -22,6 +22,11 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-grpc</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-grpc-common</artifactId>
+            <version>4.4.1-SNAPSHOT</version>
```
Comment (bulldog98) [inline integration-tests/grpc-interceptors/pom.xml:28]: Was it intended to use the snapshot?

**Comment by gsmet** [comment]
i turned it into draft. Mark it ready for review when completed.

File: integration-tests/grpc-interceptors/pom.xml (hunk 2 above)
Comment (cescoffier) [reply integration-tests/grpc-interceptors/pom.xml:28]: No that's the mistake. It's a left over of my initial tests.

**Comment by Sanne** [comment]
:+1: seems no problems with this on Hibernate Reactive, but we might do a release of it just to match.

**Comment by gsmet** [comment]
Unfortunately, some of the CI failures look related.

**Comment by Sanne** [comment]
> Unfortunately, some of the CI failures look related.

Agreed. We'll hold of from upgrading HR until it's clarified

**Comment by cescoffier** [comment]
The TCK one is annoying. @Sgitario can you help me with this one? 
I've tried to update the AsyncInvokeImpl, but while I complete the completable future with a processing exception it still sees the other exception.

**Comment by Sgitario** [comment]
I have zero experience with these tests, but I'm seeing the following exception when running the SslMutualTest.shouldFailWithNoClientSignatureCDI:

```
2023-04-11 09:04:16,629 WARN  [io.qua.boo.BootstrapAppModelFactory] (main) Expected project directory /home/jcarvaja/sources/quarkus/quarkus/tcks/microprofile-rest-client-reactive/target/quarkus-arquillian-test15827915960220448922/app does not match current project directory /home/jcarvaja/sources/quarkus/quarkus/tcks/microprofile-rest-client-reactive
2023-04-11 09:04:20,679 WARN  [io.qua.con.build] (main) The 'quarkus.rest-client-reactive.disable-smart-produces' config property is deprecated and should not be used anymore
2023-04-11 09:04:20,681 WARN  [io.qua.con.build] (main) The 'quarkus.rest-client-reactive.scope' config property is deprecated and should not be used anymore
2023-04-11 09:04:23,438 INFO  [io.quarkus] (main) quarkus-tck-microprofile-rest-client-reactive 999-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 0.884s. Listening on: http://localhost:8081
2023-04-11 09:04:23,439 INFO  [io.quarkus] (main) Profile test activated. 
2023-04-11 09:04:23,440 INFO  [io.quarkus] (main) Installed features: [cdi, rest-client-reactive, resteasy-reactive, resteasy-reactive-jsonb, servlet, smallrye-context-propagation, vertx]
2023-04-11 09:04:24,697 ERROR [io.qua.arq.QuarkusDeployableContainer] (main) Failed to copy static field: java.lang.IllegalArgumentException: Can not set static org.eclipse.microprofile.rest.client.tck.ssl.HttpsServer field org.eclipse.microprofile.rest.client.tck.ssl.AbstractSslTest.httpsServer to org.eclipse.microprofile.rest.client.tck.ssl.HttpsServer
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at java.base/jdk.internal.reflect.UnsafeStaticObjectFieldAccessorImpl.set(UnsafeStaticObjectFieldAccessorImpl.java:79)
	at java.base/java.lang.reflect.Field.set(Field.java:799)
	at io.quarkus.arquillian.QuarkusDeployableContainer.deploy(QuarkusDeployableContainer.java:213)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:151)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:118)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.executeOperation(ContainerDeployController.java:239)
	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deploy(ContainerDeployController.java:118)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

**Comment by Sgitario** [comment]
> I have zero experience with these tests, but I'm seeing the following exception when running the SslMutualTest.shouldFailWithNoClientSignatureCDI:
> 
> ```
> 2023-04-11 09:04:16,629 WARN  [io.qua.boo.BootstrapAppModelFactory] (main) Expected project directory /home/jcarvaja/sources/quarkus/quarkus/tcks/microprofile-rest-client-reactive/target/quarkus-arquillian-test15827915960220448922/app does not match current project directory /home/jcarvaja/sources/quarkus/quarkus/tcks/microprofile-rest-client-reactive
> 2023-04-11 09:04:20,679 WARN  [io.qua.con.build] (main) The 'quarkus.rest-client-reactive.disable-smart-produces' config property is deprecated and should not be used anymore
> 2023-04-11 09:04:20,681 WARN  [io.qua.con.build] (main) The 'quarkus.rest-client-reactive.scope' config property is deprecated and should not be used anymore
> 2023-04-11 09:04:23,438 INFO  [io.quarkus] (main) quarkus-tck-microprofile-rest-client-reactive 999-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 0.884s. Listening on: http://localhost:8081
> 2023-04-11 09:04:23,439 INFO  [io.quarkus] (main) Profile test activated. 
> 2023-04-11 09:04:23,440 INFO  [io.quarkus] (main) Installed features: [cdi, rest-client-reactive, resteasy-reactive, resteasy-reactive-jsonb, servlet, smallrye-context-propagation, vertx]
> 2023-04-11 09:04:24,697 ERROR [io.qua.arq.QuarkusDeployableContainer] (main) Failed to copy static field: java.lang.IllegalArgumentException: Can not set static org.eclipse.microprofile.rest.client.tck.ssl.HttpsServer field org.eclipse.microprofile.rest.client.tck.ssl.AbstractSslTest.httpsServer to org.eclipse.microprofile.rest.client.tck.ssl.HttpsServer
> 	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
> 	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
> 	at java.base/jdk.internal.reflect.UnsafeStaticObjectFieldAccessorImpl.set(UnsafeStaticObjectFieldAccessorImpl.java:79)
> 	at java.base/java.lang.reflect.Field.set(Field.java:799)
> 	at io.quarkus.arquillian.QuarkusDeployableContainer.deploy(QuarkusDeployableContainer.java:213)
> 	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:151)
> 	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController$3.call(ContainerDeployController.java:118)
> 	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.executeOperation(ContainerDeployController.java:239)
> 	at org.jboss.arquillian.container.impl.client.container.ContainerDeployController.deploy(ContainerDeployController.java:118)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> ```

Note that this error seems unrelated (though it should be fixed in the Arquillian extension side).
I'm trying to see why Vert.x was returning the expected IOException when jetty returned `javax.net.ssl.SSLHandshakeException: Empty client certificate chain`, and now it returns `io.vertx.core.http.HttpClosedException: Connection was closed`.

**Comment by Sgitario** [comment]
The root cause of this issue seems to be that Netty is not throwing the handshake failure exception. This was done in [here](https://github.com/netty/netty/blob/24a0ac36ea91d1aee647d738f879ac873892d829/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1894).

To be precise, before these changes, the condition `handshakePromise.tryFailure(cause)` returned true and hence the exception was thrown, but after these changes, it returns false. This logic that returns either false or true is: https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultPromise.java#L632

This logic has not changed in ages, so it must be something else that causes the change of behaviour (not sure if this is caused by a change in Vert.x or Netty). So, we should ask for support first from somebody in their teams.

**Comment by Sanne** [comment]
> The root cause of this issue seems to be that Netty is not throwing the handshake failure exception. This was done in [here](https://github.com/netty/netty/blob/24a0ac36ea91d1aee647d738f879ac873892d829/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1894).
> 
> To be precise, before these changes, the condition `handshakePromise.tryFailure(cause)` returned true and hence the exception was thrown, but after these changes, it returns false. This logic that returns either false or true is: https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultPromise.java#L632
> 
> This logic has not changed in ages, so it must be something else that causes the change of behaviour (not sure if this is caused by a change in Vert.x or Netty). So, we should ask for support first from somebody in their teams.

@franz1981 do you think you could help here?

**Comment by franz1981** [comment]
If the result is already set, there is not much that can be done (it returns true when is already set): we should Trak which changes already make it to be set or, I'd return false, what's already set to handle it right

**Comment by cescoffier** [comment]
This PR updates to netty .90. Maybe we can skip this update and stay on 89 for now.

**Comment by franz1981** [comment]
> This PR updates to netty .90. Maybe we can skip this update and stay on 89

89 isn't a good idea, see the regressions list at https://netty.io/news/2023/03/14/4-1-90-Final.html

**Comment by tsegismont** [comment]
@cescoffier @gsmet https://github.com/cescoffier/quarkus/pull/196

**Comment by gsmet** [comment]
Thanks @tsegismont . So that lets us with the TCK issues.

**Comment by cescoffier** [comment]
@Sgitario What's the status about the TCK issues? (maybe we should do a short call about them)

**Comment by Sgitario** [comment]
> @Sgitario What's the status about the TCK issues? (maybe we should do a short call about them)

My comment https://github.com/quarkusio/quarkus/pull/32416#issuecomment-1502976982 still stands. 
From the comment https://github.com/quarkusio/quarkus/pull/32416#issuecomment-1503022289, it seems that the result is being set, and since this is used by Vert.x, I assume that there is a regression issue on the Vert.x side.

**Comment by cescoffier** [comment]
I got a fix, not clean but seems to work.

It's a change in Vert.x but that change makes sense. It's then a rewrapping game in the rest client and the proxy that caused the issue.

**Comment by Sgitario** [comment]
> I got a fix, not clean but seems to work.
> 
> It's a change in Vert.x but that change makes sense. It's then a rewrapping game in the rest client and the proxy that caused the issue.

Really nice catchup @cescoffier ! https://github.com/quarkusio/quarkus/commit/5e8c6bffb07313ab0c9579e07d32c66dbdd4d097

**Comment by gsmet** [comment]
There's a problem with the PR. It includes way too many commits.

**Comment by cescoffier** [comment]
@gsmet oups! I will fix that after lunch.

**Comment by cescoffier** [comment]
Ok, single commit - all included.

**Review by geoand — APPROVED** [review]
🎉

**Comment by Sanne** [comment]
Nice! we'll upgrade Hibernate Reactive as well then.

**Pull request merged**